diff -u -r ceph-0.80.1/src/init-ceph.in ceph-0.80.1.patched/src/init-ceph.in
--- ceph-0.80.1/src/init-ceph.in	2014-05-12 17:14:26.000000000 +0200
+++ ceph-0.80.1.patched/src/init-ceph.in	2014-05-13 17:57:32.997382297 +0200
@@ -4,7 +4,7 @@
 
 ### BEGIN INIT INFO
 # Provides:          ceph
-# Default-Start:     2 3 4 5
+# Default-Start:     3 5
 # Default-Stop:      0 1 6
 # Required-Start:    $remote_fs $named $network $time
 # Required-Stop:     $remote_fs $named $network $time
diff -u -r ceph-0.80.1/src/init-radosgw ceph-0.80.1.patched/src/init-radosgw
--- ceph-0.80.1/src/init-radosgw	2014-05-12 17:14:25.000000000 +0200
+++ ceph-0.80.1.patched/src/init-radosgw	2014-05-13 17:58:10.525298680 +0200
@@ -3,7 +3,7 @@
 # Provides:          radosgw
 # Required-Start:    $remote_fs $named $network $time
 # Required-Stop:     $remote_fs $named $network $time
-# Default-Start:     2 3 4 5
+# Default-Start:     3 5
 # Default-Stop:      0 1 6
 # Short-Description: radosgw RESTful rados gateway
 # Description: radosgw RESTful rados gateway
