diff --git a/src/common/sync_filesystem.h b/src/common/sync_filesystem.h
index dc90b89..d7523db 100644
--- a/src/common/sync_filesystem.h
+++ b/src/common/sync_filesystem.h
@@ -25,6 +25,18 @@
 # endif
 #endif
 
+#ifndef __NR_syncfs
+# if defined(__i386__)
+#  define __NR_syncfs 344
+#  define USE_SYNCFS_NR
+# elif defined(__x86_64__)
+#  define __NR_syncfs 306
+#  define USE_SYNCFS_NR
+# else
+#  error please define __NR_syncfs for your architecture
+# endif
+#endif
+
 inline int sync_filesystem(int fd)
 {
   /* On Linux, newer versions of glibc have a function called syncfs that
