Index: ceph-0.80.1/src/ceph-disk
===================================================================
--- ceph-0.80.1.orig/src/ceph-disk
+++ ceph-0.80.1/src/ceph-disk
@@ -51,7 +51,7 @@ Activate:
  - triggered by admin 'ceph-disk activate <path>'
  - triggered on ceph service startup with 'ceph-disk activate-all'
 
-We rely on /dev/disk/by-partuuid to find partitions by their UUID;
+We rely on /dev/disk/by-uuid to find partitions by their UUID;
 this is what the journal symlink inside the osd data volume normally
 points to.
 
@@ -63,7 +63,7 @@ udev triggers 'ceph-disk activate <dev>'
 activate-journal <dev>' based on the partition type.
 
 On old distros (e.g., RHEL6), the blkid installed does not recognized
-GPT partition metadata and the /dev/disk/by-partuuid etc. links aren't
+GPT partition metadata and the /dev/disk/by-uuid etc. links aren't
 present.  We have a horrible hack in the form of ceph-disk-udev that
 parses gparted output to create the symlinks above and triggers the
 'ceph-disk activate' etc commands that udev normally would do if it
@@ -952,6 +952,10 @@ def prepare_journal_dev(
             num=num,
             size=journal_size,
             )
+        journal_symlink = '{data}{num}'.format(
+            data=data,
+            num=num,
+            )
     else:
         # sgdisk has no way for me to say "whatever is the next
         # free index number" when setting type guids etc, so we
@@ -964,6 +968,9 @@ def prepare_journal_dev(
             num=num,
             size=journal_size,
             )
+        journal_symlink = '/dev/disk/by-uuid/{journal_uuid}'.format(
+            journal_uuid=journal_uuid,
+            )
         LOG.warning('OSD will not be hot-swappable if journal is not the same device as the osd data')
 
     try:
@@ -1022,10 +1029,6 @@ def prepare_journal_dev(
                 ],
             )
 
-        journal_symlink = '/dev/disk/by-partuuid/{journal_uuid}'.format(
-            journal_uuid=journal_uuid,
-            )
-
         journal_dmcrypt = None
         if journal_dm_keypath:
             journal_dmcrypt = journal_symlink
@@ -1982,7 +1985,7 @@ def main_activate_journal(args):
     activate_lock.acquire()
     try:
         osd_uuid = get_journal_osd_uuid(args.dev)
-        path = os.path.join('/dev/disk/by-partuuid/', osd_uuid.lower())
+        path = os.path.join('/dev/disk/by-uuid/', osd_uuid.lower())
 
         (cluster, osd_id) = mount_activate(
             dev=path,
