diff -r -u ceph-0.80.orig/src/init-ceph.in ceph-0.80/src/init-ceph.in
--- ceph-0.80.orig/src/init-ceph.in	2014-04-28 14:43:07.382157772 +0200
+++ ceph-0.80/src/init-ceph.in	2014-04-28 14:44:16.552987203 +0200
@@ -4,7 +4,7 @@
 
 ### BEGIN INIT INFO
 # Provides:          ceph
-# Default-Start:     2 3 4 5
+# Default-Start:     3 4 5
 # Default-Stop:      0 1 6
 # Required-Start:    $remote_fs $named $network $time
 # Required-Stop:     $remote_fs $named $network $time
diff -r -u ceph-0.80.bak/src/init-radosgw ceph-0.80/src/init-radosgw
--- ceph-0.80.bak/src/init-radosgw	2014-04-28 14:43:07.386157760 +0200
+++ ceph-0.80/src/init-radosgw	2014-04-28 14:44:53.296896599 +0200
@@ -3,7 +3,7 @@
 # Provides:          radosgw
 # Required-Start:    $remote_fs $named $network $time
 # Required-Stop:     $remote_fs $named $network $time
-# Default-Start:     2 3 4 5
+# Default-Start:     3 4 5
 # Default-Stop:      0 1 6
 # Short-Description: radosgw RESTful rados gateway
 # Description: radosgw RESTful rados gateway
