diff -r -u ceph-0.80.1/Makefile.am ceph-0.80.1.orignial/Makefile.am
--- ceph-0.80.1/Makefile.am	2014-05-27 16:10:14.541274001 +0200
+++ ceph-0.80.1.orignial/Makefile.am	2014-05-27 15:10:16.875115539 +0200
@@ -24,13 +24,6 @@
 	-install -m 644 share/id_dsa_drop.ceph.com $(DESTDIR)$(datadir)/ceph/id_dsa_drop.ceph.com
 	-install -m 644 share/id_dsa_drop.ceph.com.pub $(DESTDIR)$(datadir)/ceph/id_dsa_drop.ceph.com.pub
 
-# why is it so hard to make autotools to this?
-install-data-local:
-	-mkdir -p $(DESTDIR)$(datadir)/ceph
-	-install -m 644 share/known_hosts_drop.ceph.com $(DESTDIR)$(datadir)/ceph/known_hosts_drop.ceph.com
-	-install -m 644 share/known_hosts_drop.ceph.com $(DESTDIR)$(datadir)/ceph/id_dsa_drop.ceph.com
-	-install -m 644 share/known_hosts_drop.ceph.com $(DESTDIR)$(datadir)/ceph/id_dsa_drop.ceph.com.pub
-
 all-local:
 if WITH_DEBUG
 #	We need gtest to build the rados-api tests. We only build those in
