diff -up ceph-0.46/src/init-ceph.in.orig ceph-0.46/src/init-ceph.in
--- ceph-0.46/src/init-ceph.in.orig	2012-04-24 16:06:39.000000000 -0400
+++ ceph-0.46/src/init-ceph.in	2012-05-09 16:41:36.500004052 -0400
@@ -1,11 +1,11 @@
 #!/bin/sh
 # Start/stop ceph daemons
-# chkconfig: 2345 60 80
+# chkconfig: - 60 80

 ### BEGIN INIT INFO
 # Provides:          ceph
-# Default-Start:     2 3 4 5
-# Default-Stop:      0 1 6
+# Default-Start:
+# Default-Stop:
 # Required-Start:    $remote_fs $named $network $time
 # Required-Stop:     $remote_fs $named $network $time
 # Short-Description: Start Ceph distributed file system daemons at boot time
