Index: configure.in
===================================================================
--- configure.in	(revision 29691)
+++ configure.in	(working copy)
@@ -1742,6 +1742,7 @@
   packaging/svr4/pkginfo
   plugins/Makefile
   plugins/asn1/Makefile
+  plugins/ceph/Makefile
   plugins/docsis/Makefile
   plugins/ethercat/Makefile
   plugins/giop/Makefile
Index: Makefile.am
===================================================================
--- Makefile.am	(revision 29691)
+++ Makefile.am	(working copy)
@@ -264,6 +264,7 @@
 -include plugins/Custom.make
 plugin_ldadd = $(_CUSTOM_plugin_ldadd_) \
 	-dlopen plugins/asn1/asn1.la \
+	-dlopen plugins/ceph/ceph.la \
 	-dlopen plugins/docsis/docsis.la \
 	-dlopen plugins/ethercat/ethercat.la \
 	-dlopen plugins/giop/cosnaming.la \
Index: plugins/Makefile.nmake
===================================================================
--- plugins/Makefile.nmake	(revision 29691)
+++ plugins/Makefile.nmake	(working copy)
@@ -33,6 +33,9 @@
 	cd asn1
 	$(MAKE) /$(MAKEFLAGS) -f Makefile.nmake $(PLUGIN_TARGET)
 	cd ..
+	cd ceph
+	$(MAKE) /$(MAKEFLAGS) -f Makefile.nmake $(PLUGIN_TARGET)
+	cd ..
 	cd docsis
 	$(MAKE) /$(MAKEFLAGS) -f Makefile.nmake $(PLUGIN_TARGET)
 	cd ..
@@ -86,6 +89,7 @@
 !IFDEF ENABLE_LIBWIRESHARK
 	cd..
 	xcopy plugins\asn1\*.dll $(INSTALL_DIR)\plugins\$(VERSION) /d
+	xcopy plugins\ceph\*.dll $(INSTALL_DIR)\plugins\$(VERSION) /d
 	xcopy plugins\docsis\*.dll $(INSTALL_DIR)\plugins\$(VERSION) /d
 	xcopy plugins\ethercat\*.dll $(INSTALL_DIR)\plugins\$(VERSION) /d
 	xcopy plugins\giop\*.dll $(INSTALL_DIR)\plugins\$(VERSION) /d
Index: plugins/Makefile.am
===================================================================
--- plugins/Makefile.am	(revision 29691)
+++ plugins/Makefile.am	(working copy)
@@ -24,6 +24,7 @@
 -include Custom.make
 SUBDIRS = $(_CUSTOM_SUBDIRS_) \
 	asn1 \
+	ceph \
 	docsis \
 	ethercat \
 	giop \
Index: epan/Makefile.am
===================================================================
--- epan/Makefile.am	(revision 29691)
+++ epan/Makefile.am	(working copy)
@@ -201,6 +201,7 @@
 -include ../plugins/Custom.make
 plugin_src = \
 	../plugins/asn1/packet-asn1.c \
+	../plugins/ceph/packet-ceph.c \
 	../plugins/docsis/packet-bintrngreq.c \
 	../plugins/docsis/packet-bpkmattr.c \
 	../plugins/docsis/packet-bpkmreq.c \
Index: packaging/nsis/Makefile.nmake
===================================================================
--- packaging/nsis/Makefile.nmake	(revision 29691)
+++ packaging/nsis/Makefile.nmake	(working copy)
@@ -44,6 +44,7 @@
 
 PLUGINS= \
 	../../plugins/asn1/asn1.dll \
+	../../plugins/ceph/ceph.dll \
 	../../plugins/docsis/docsis.dll \
 	../../plugins/ethercat/ethercat.dll \
 	../../plugins/giop/coseventcomm.dll \
Index: packaging/nsis/wireshark.nsi
===================================================================
--- packaging/nsis/wireshark.nsi	(revision 29691)
+++ packaging/nsis/wireshark.nsi	(working copy)
@@ -875,6 +875,7 @@
 ;-------------------------------------------
 SetOutPath '$INSTDIR\plugins\${VERSION}'
 File "..\..\plugins\asn1\asn1.dll"
+File "..\..\plugins\ceph\ceph.dll"
 File "..\..\plugins\docsis\docsis.dll"
 File "..\..\plugins\ethercat\ethercat.dll"
 File "..\..\plugins\giop\coseventcomm.dll"
