Index: configure.in
===================================================================
--- configure.in	(revision 26859)
+++ configure.in	(working copy)
@@ -1671,6 +1671,7 @@
   plugins/agentx/Makefile
   plugins/artnet/Makefile
   plugins/asn1/Makefile
+  plugins/ceph/Makefile
   plugins/ciscosm/Makefile
   plugins/docsis/Makefile
   plugins/enttec/Makefile
Index: Makefile.am
===================================================================
--- Makefile.am	(revision 26859)
+++ Makefile.am	(working copy)
@@ -243,6 +243,7 @@
 	-dlopen plugins/agentx/agentx.la \
 	-dlopen plugins/artnet/artnet.la \
 	-dlopen plugins/asn1/asn1.la \
+	-dlopen plugins/ceph/ceph.la \
 	-dlopen plugins/ciscosm/ciscosm.la \
 	-dlopen plugins/docsis/docsis.la \
 	-dlopen plugins/enttec/enttec.la \
Index: plugins/Makefile.nmake
===================================================================
--- plugins/Makefile.nmake	(revision 26859)
+++ plugins/Makefile.nmake	(working copy)
@@ -39,6 +39,9 @@
 	cd asn1
 	$(MAKE) /$(MAKEFLAGS) -f Makefile.nmake $(PLUGIN_TARGET)
 	cd ..
+	cd ceph
+	$(MAKE) /$(MAKEFLAGS) -f Makefile.nmake $(PLUGIN_TARGET)
+	cd ..
 	cd ciscosm
 	$(MAKE) /$(MAKEFLAGS) -f Makefile.nmake $(PLUGIN_TARGET)
 	cd ..
@@ -128,6 +131,7 @@
 	xcopy agentx\*.dll $(VERSION) /d
 	xcopy artnet\*.dll $(VERSION) /d
 	xcopy asn1\*.dll $(VERSION) /d
+	xcopy ceph\*.dll $(VERSION) /d
 	xcopy ciscosm\*.dll $(VERSION) /d
 	xcopy docsis\*.dll $(VERSION) /d
 	xcopy enttec\*.dll $(VERSION) /d
Index: plugins/Makefile.am
===================================================================
--- plugins/Makefile.am	(revision 26859)
+++ plugins/Makefile.am	(working copy)
@@ -26,6 +26,7 @@
 	agentx \
 	artnet \
 	asn1 \
+	ceph \
 	ciscosm \
 	docsis \
 	enttec \
Index: epan/Makefile.am
===================================================================
--- epan/Makefile.am	(revision 26859)
+++ epan/Makefile.am	(working copy)
@@ -189,6 +189,7 @@
 plugin_src = \
 	../plugins/artnet/packet-artnet.c \
 	../plugins/asn1/packet-asn1.c \
+	../plugins/ceph/packet-ceph.c \
 	../plugins/docsis/packet-bpkmattr.c \
 	../plugins/docsis/packet-bpkmreq.c \
 	../plugins/docsis/packet-bpkmrsp.c \
Index: packaging/nsis/Makefile.nmake
===================================================================
--- packaging/nsis/Makefile.nmake	(revision 26859)
+++ packaging/nsis/Makefile.nmake	(working copy)
@@ -47,6 +47,7 @@
 	../../plugins/agentx/agentx.dll \
 	../../plugins/artnet/artnet.dll \
 	../../plugins/asn1/asn1.dll \
+	../../plugins/ceph/ceph.dll \
 	../../plugins/ciscosm/ciscosm.dll \
 	../../plugins/docsis/docsis.dll \
 	../../plugins/enttec/enttec.dll \
Index: packaging/nsis/wireshark.nsi
===================================================================
--- packaging/nsis/wireshark.nsi	(revision 26859)
+++ packaging/nsis/wireshark.nsi	(working copy)
@@ -779,6 +779,7 @@
 File "..\..\plugins\agentx\agentx.dll"
 File "..\..\plugins\artnet\artnet.dll"
 File "..\..\plugins\asn1\asn1.dll"
+File "..\..\plugins\ceph\ceph.dll"
 File "..\..\plugins\ciscosm\ciscosm.dll"
 File "..\..\plugins\docsis\docsis.dll"
 File "..\..\plugins\enttec\enttec.dll"
